src/malli_ts/data_mapping.cljc

```diff
@@ -0,0 +1,210 @@
+
+       (and registry get-schema-name)
+       , (let [get-schema-name (if (fn? get-schema-name)
+                                 get-schema-name
+                                 (fn to-clj-get-schema-nm [o]
+                                   (let [schema-nm (if (keyword? get-schema-name)
+                                                     (name get-schema-name)
+                                                     (str get-schema-name))]
+                                     (js/goog.object.get o schema-nm nil))))
+               obj             (if (and (array? data)
+                                        (>= (js/goog.object.get data "length" 0) 1))
+                                 (aget data 0)
+                                 data)
+               clj<->js-map    (clj<->js-key-mapping-cached registry
+                                                            (get-schema-name obj))]
+           (to-clj' data clj<->js-map))
+
+       :else
+       , (to-clj data :registry registry :get-schema-name default-js-get-schema-name)))
+
+   (declare map-proxy)
+
+   (defn- array-push
+     ([res] res)
+     ([res x] (doto res (.push x))))
+
+   (defn into-js-array
+     [xform from]
+     (transduce xform array-push (array) from))
+
+   (defn- to-js'
+     [data js-props->clj-keys]
+     (when js-props->clj-keys
+       (cond
+         (or (sequential? data)
+             (set? data))
+         (into-js-array (map #(to-js' % js-props->clj-keys)) data)
+
+         (associative? data)
+         (map-proxy data js-props->clj-keys)
+
+         :else
+         data)))
+
+   (defn ^:export to-js
+     [data & {:keys [registry get-schema-name]
+              :as   schema}]
+     (cond
+       (m/schema? schema)
+       , (let [{:keys [js-props->clj-keys]} (clj<->js-key-mapping-cached
+                                             schema)]
+           (to-js' data js-props->clj-keys))
+
+       (and registry get-schema-name)
+       , (let [is-coll   (or (sequential? data)
+                             (set? data))
+               schema-nm (if is-coll
+                           (-> data
+                               first
+                               get-schema-name)
+                           (get-schema-name data))
+               {:keys [js-props->clj-keys]}
+               (clj<->js-key-mapping-cached registry schema-nm)]
+           (to-js' data js-props->clj-keys))
+
+       :else
+       , (to-js data :registry registry :get-schema-name default-get-schema-name)))
+
+   (defn- map-proxy-get
+     [js-props->clj-keys target key]
+     (case key
+       "unwrap/clj" target
+
+       (-> js-props->clj-keys
+           (get key)
+           (as-> k (get target k))
+           (to-js' js-props->clj-keys))))
+
+   (defn- map-proxy
+     [data js-props->clj-keys]
+     (if (instance? JsProxy data)
+       data
+       (js/Proxy. data
+                  #js
+                  {:get            (partial map-proxy-get js-props->clj-keys)
```

Comment:
@flow-danny yes, `js-props->clj-keys` is flat per root schema. One of the reasons is because `bean`'s `:prop->key` and `:key->prop` only gives you a key/property name, but it doesn't tell you where in the hierarchy you are, for example by giving you a path (in the case of nested maps). So you're forced to look the mapping up in a flat index for all keys/props across the root schema/map's keys/properties.